docs/contribute/addnode.qmd

```diff
@@ -0,0 +1,170 @@
+from pydantic import baseModel
+
+from ribasim.input_base import InputMixin
+
+__all__ = ("NewNodeType",)
+
+class StaticSchema(pa.SchemaModel):
+    node_id: Series[int] = pa.Field(coerce=True)
+    # Series per node property, pa.Field for mandatory data
+
+# Possible other schemas
+
+
+class NewNodeType(InputMixin, BaseModel):
+    """
+    Description of this node type.
+
+    Parameters
+    ----------
+    static: pandas.DataFrame
+        table with data for this node type.
+
+    possible other schemas
+    """
+
+    _input_type = "NewNodeType"
+    static: Optional[DataFrame[StaticSchema]] = None
+    # possible other schemas
+
+    class Config:
+        validate_assignment = True
+
+```
+
+in `python/ribasim/ribasim/model.py`, add
+
+- `from ribasim.new_node_type import NewNodeType`;
+- `(NewNodetype, "new_node_type")` in the `_NODES` list;
+- new_node_type as a parameter and in the docstring of the `Model` class.
+
+
+# QGIS plugin
+
+The script `qgis\core\nodes.py` has to be updated to specify how the new node type is displayed by the QGIS plugin. Specifically:
+
+- Add `"NewNodeType: NewNodeType",` to the `"map"` dictionary in `Node.set_editor_widget`; 
+- Add a color and shape description in the `markers` dictionary in `Node.renderer`.
+- Add an input class per schema, e.g. 
+
+```python
+class NewNodeTypeStatic
+    input_type = "NewNodeType / static"
+    geometry_type = "No Geometry"
+    attributes = [
+        QgsField("node_id", QVariant.Int)
+        # Other fields for properties of this node
+    ]
+```
+
+and add these to the NODES dictionary.
+
+
+# Tests
+
+Models for the julia tests are generated by running `python/ribasim/tests/conftest.py`. These models should also be updated to contain the new node type. Note that certain tests must be updated accordingly when the models used for certain tests are updated, e.g. the final state of the models in `core/test/basin.jl`.
+
+Future:
+- Test for the specific behaviour of each node type
+
+
+# Documentation
+
+There are several parts of the documentation which should be updated with the new node type:
+
+- `docs/core/equations` should contain a short explanation and if possible an analytical expression for the behaviour of the new node;
+- `docs\core\usage.qmd` should conntain a short explanation of the node and the possible schemas associated with it;
+- The example models constructed in `docs/python/examples.ipynb` should be extended with the new node type or a new example model with the new node type should be made.
+
+# Finishing up 
+
+When a new node type is created, one needs to run
+
+```
+julia --project=docs docs/gen_schema.jl
+```
+to create new jsonschemas in the docs folder, and
+
+```
+datamodel-codegen --use-title-as-name --input docs/schema/root.schema.json --output python/ribasim/ribasim/models.py
+```
+to generate the Python models from it.
+
+This should be followed by `black python` to format the generated code.
```

Comment:
Perhaps running `black python` for formatting and `ruff python` for linting should be added to the Python tooling development page.

Then we can link to the sections that a user should perform at this stage;

1. running black and ruff
2. running `pytest python/ribasim` (shall we update https://deltares.github.io/Ribasim/contribute/python.html#running-the-tests directly to use the repo root as working directory?)
3. running the julia core tests